torch/distributed/optim/optimizer.py

```diff
@@ -0,0 +1,112 @@
+import torch.distributed.rpc as rpc
+import torch.distributed.autograd as dist_autograd
+
+from collections import defaultdict
+
+class FunctionalOptimizer:
+    """
+    Base class for functional optimizers.
+
+    Functional optimizers are similar to torch.optim optimizers, but
+    instead of using params.grad as input, it takes the gradients as input to
+    the step function. This allows to implement multi-threaded versions of
+    such as Hogwild!.
+
+    Args:
+        params (list): list of parameters to optimize. order of the list
+                       matters as the list of gradients passed to the step
+                       function must be aligned with this list.
+    """
+    def __init__(self, params):
+        self.params = params
+
+    def step(self, gradients):
```

Comment:
The current `Optimizer` takes and returns the loss function. What would be an equivalent here?

Reply:
Unfortunately there's no straightforward way of porting the closure() functionality to distributed as it would potentially require algorithm-specific synchronization across workers. This would be true for L-FBGS for example, where closure() and parameter updates are interleaved in a loop inside of step().